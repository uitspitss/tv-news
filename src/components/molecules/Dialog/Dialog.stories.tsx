import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';

import { Dialog, DialogProps } from './Dialog';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: Dialog,
} as Meta;

const Template: Story<DialogProps> = (args) => <Dialog {...args} />;

export const Default = Template.bind({});
Default.args = {
  trigger: <button type="button">click me</button>,
  children: (close) => (
    <div>
      <div>content</div>
      <div>
        <button type="button" onClick={close}>
          close
        </button>
      </div>
    </div>
  ),
  overlay: true,
};
