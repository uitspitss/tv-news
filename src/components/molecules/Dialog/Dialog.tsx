import React, { FC, memo, ReactNode } from 'react';
import Popup from 'reactjs-popup';
import { PopupActions } from 'reactjs-popup/dist/types';
// eslint-disable-next-line import/no-extraneous-dependencies
import tw, { styled } from 'twin.macro';

type DialogCloseAction = PopupActions['close'];

export type { DialogCloseAction };

type BaseProps = {
  /**
   * class for twin.macro
   */
  className?: string;
  /**
   * children（required）
   */
  children: (close: DialogCloseAction) => ReactNode;
  /**
   * overlay
   */
  overlay?: boolean;
};

export type DialogProps =
  | (BaseProps & {
      //  eslint-disable-next-line no-undef
      trigger: JSX.Element;
    })
  | (BaseProps & {
      open: boolean;
      onClose: () => void;
    });

type Props = DialogProps;

const Component: FC<Props> = (props) => {
  const { className, children, ...rest } = props;

  return (
    <Popup className={className} arrow={false} modal {...rest}>
      {(close: DialogCloseAction) => (
        <div className="dialog">{children(close)}</div>
      )}
    </Popup>
  );
};

const StyledComponent = styled(Component)`
  &-overlay {
    ${({ overlay }) => overlay && tw`bg-gray-500 bg-opacity-50`}
  }

  &-content {
    ${tw`bg-white rounded p-4`}
  }
`;

/**
 * ダイアログ
 */
export const Dialog: FC<DialogProps> = memo((props) => (
  <StyledComponent {...props} />
));
