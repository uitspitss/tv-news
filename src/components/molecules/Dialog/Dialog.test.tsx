import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { Dialog } from './Dialog';

describe('molecules/Dialog', () => {
  test('snapshot test', () => {
    const Component = () => (
      <Dialog trigger={<button type="button">trigger</button>}>
        {(close) => (
          <div>
            <div>content</div>
            <div>
              <button type="button" onClick={close}>
                close
              </button>
            </div>
          </div>
        )}
      </Dialog>
    );
    const { container } = render(<Component />);
    expect(container).toMatchSnapshot();
  });

  describe('Behavior', () => {
    const Component = () => (
      <Dialog trigger={<button type="button">trigger</button>}>
        {(close) => (
          <div>
            <div>content</div>
            <div>
              <button type="button" onClick={close}>
                close
              </button>
            </div>
          </div>
        )}
      </Dialog>
    );

    beforeEach(() => {
      jest.clearAllMocks();
      render(<Component />);
    });

    test('initial view', () => {
      expect(
        screen.getByRole('button', { name: 'trigger' }),
      ).toBeInTheDocument();
    });

    test('click, then show content', () => {
      fireEvent.click(screen.getByRole('button', { name: 'trigger' }));

      expect(screen.getByText('content')).toBeInTheDocument();
    });

    test('close button click, then closed,', () => {
      fireEvent.click(screen.getByRole('button', { name: 'trigger' }));

      expect(screen.getByText('content')).toBeInTheDocument();

      fireEvent.click(screen.getByRole('button', { name: 'close' }));

      expect(screen.queryByText('content')).not.toBeInTheDocument();
    });
  });
});
