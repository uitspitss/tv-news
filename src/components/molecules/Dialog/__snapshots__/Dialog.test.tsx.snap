// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecules/Dialog snapshot test 1`] = `
<div>
  <button
    aria-describedby="popup-1"
    type="button"
  >
    trigger
  </button>
</div>
`;
